persistence-migration/src/main/resources/migration/changelog-v5.6.0.xml

```diff
@@ -2353,4 +2353,255 @@
+                    foreignKeyName="ROLES_PERMISSIONS_ROLE_FK"
+                    referencedTableName="ROLE" referencedColumnNames="ID"
+                    deferrable="true" initiallyDeferred="true" deleteCascade="true"
+                    validatePrimaryKey="true" validateForeignKey="true" />
+            </column>
+
+            <column name="PERMISSION_ID" type="BIGINT">
+                <constraints nullable="false" primaryKey="true"
+                    primaryKeyName="ROLES_PERMISSIONS_PK"
+                    foreignKeyName="ROLES_PERMISSIONS_PERMISSION_FK"
+                    referencedTableName="PERMISSION" referencedColumnNames="ID"
+                    deferrable="true" initiallyDeferred="true" deleteCascade="true"
+                    validatePrimaryKey="true" validateForeignKey="true" />
+            </column>
+        </createTable>
+
+        <createTable ifNotExists="true" tableName="USER_PROJECT_ROLES">
+            <column autoIncrement="true" name="ID" type="BIGINT">
+                <constraints nullable="false" primaryKey="true" primaryKeyName="USER_PROJECT_ROLES_PK" />
+            </column>
+
+            <column name="USER_ID" type="BIGINT">
+                <constraints nullable="false"
+                    foreignKeyName="USER_PROJECT_ROLES_USER_FK"
+                    referencedTableName="USER" referencedColumnNames="ID"
+                    deferrable="true" initiallyDeferred="true" deleteCascade="true"
+                    validateForeignKey="true" />
+            </column>
+
+            <column name="PROJECT_ID" type="BIGINT">
+                <constraints nullable="false"
+                    foreignKeyName="USER_PROJECT_ROLES_PROJECT_FK"
+                    referencedTableName="PROJECT" referencedColumnNames="ID"
+                    deferrable="true" initiallyDeferred="true" deleteCascade="true"
+                    validateForeignKey="true" />
+            </column>
+
+            <column name="ROLE_ID" type="BIGINT">
+                <constraints nullable="false"
+                    foreignKeyName="USER_PROJECT_ROLES_ROLE_FK"
+                    referencedTableName="ROLE" referencedColumnNames="ID"
+                    deferrable="true" initiallyDeferred="true" deleteCascade="true"
+                    validateForeignKey="true" />
+            </column>
+        </createTable>
+
+        <createIndex unique="true" tableName="USER_PROJECT_ROLES" indexName="USER_PROJECT_ROLES_IDX">
+            <column name="USER_ID" />
+            <column name="PROJECT_ID" />
+            <column name="ROLE_ID" />
+        </createIndex>
+
+        <sql splitStatements="false">
+            -- Helper function to recalculate all user permissions for a project.
+            -- Called by trigger functions to update the values in the USER_PROJECT_EFFECTIVE_PERMISSIONS table.
+            CREATE OR REPLACE FUNCTION recalc_user_project_role_effective_permissions(project_ids BIGINT[])
```

Comment:
We will need to clean up triggers and functions we created earlier, which are based on `PROJECT_ACCESS_TEAMS` rather than roles:

* `recalc_user_project_effective_permissions`, which is called by:
    * `effective_permissions_mx_on_delete`
    * `effective_permissions_mx_on_insert`
    * `effective_permissions_mx_on_update`
* The triggers calling the functions above:
    * `"USERS_TEAMS".trigger_effective_permissions_mx_on_users_teams_delete`
    * `"USERS_TEAMS".trigger_effective_permissions_mx_on_users_teams_insert`
    * `"USERS_TEAMS".trigger_effective_permissions_mx_on_users_teams_update`

Otherwise we'd have multiple triggers competing to populate the `USER_PROJECT_EFFECTIVE_PERMISSIONS` table on different occasions, using different data.